File: Java Blind 75/Tree/Binary Tree Level Order Traversal.java
package Tree;
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        //use a queue to explore every children (BFS)
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        while (!queue.isEmpty()) {
            //store queue size because the size is going to 
            //change during the for loop
            int qLen = queue.size(); 
            List<Integer> list = new ArrayList<>();
            for (int i = 0; i < qLen; i++) {
                //grab the node that's in front of the queue
                TreeNode currNode = queue.poll();
                if (currNode != null) {
                    //add the current node's children to the 
                    //queue to be explored later 
                    queue.add(currNode.left);
                    queue.add(currNode.right);
                    list.add(currNode.val);
                }
            }
            //don't want add an empty list to the answer
            //linked list so this is a check
            if (!list.isEmpty()) ans.add(list);
        }
        return ans;
    }
}

File: Python Blind 75/Arrays/Binary Tree Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def levelOrder(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: List[List[int]]
        """
        # Time: O(n), Space: O(n)
        res = []

        q = collections.deque()
        q.append(root) # start the queue with the root

        while q: # queue is not empty
            qLen = len(q) # counts each node in the level
            level = []
            for i in range(qLen): # go through each node in the level
                currNode = q.popleft()
                if currNode: # currNode could be null and we wouldn't want to append a null node
                    q.append(currNode.left)
                    q.append(currNode.right)
                    level.append(currNode.val)
            if level: # level could be empty and we don't want to append an empty list
                res.append(level)
        return res
